echopop/survey.py

```diff
@@ -1230,3 +1246,238 @@ def krige( self ,
                 'kriged_biomass_df': kriged_dataframe
             }
         )
+        
+        ### TODO: This should be refactored out as an external function 
+        ###### rather than a Survey-class method.
+        ### Apportion biomass based on age and sex
+        self.apportion_kriged_biomass( species_id )
+
+    def apportion_kriged_biomass( self ,
+                                  species_id ):
+        """
+        Apportion kriged biomass based on sex, age, and other proportions
+        
+        Notes
+        -----
+        This function apportions the kriged biomass values (interpolated along
+        a user-defined mesh) based on sex and age. This provides biomass estimates
+        for aged and unaged fish, total biomass inferred from the biological and 
+        kriged bioamss estimates, the coefficient of variation within each cell, 
+        and georeferenced biomass estimates for specific paired values of sex 
+        and age-class.
+        """  
+        ### Length data (Station 1 - Unaged - Sexxed )
+        length_df_copy = self.biology[ 'length_df' ].copy( ).pipe( lambda df: df.loc[ df.species_id == species_id ] )
+
+        ### Length-weight data (Station 2 - Aged - Sexed)
+        specimen_df_copy = self.biology[ 'specimen_df' ].copy( ).pipe( lambda df: df.loc[ df.species_id == species_id ] )
+
+        ### Haul data 
+        haul_catch_df = self.biology[ 'catch_df' ].copy( ).pipe( lambda df: df.loc[ df.species_id == species_id ] )
+
+        ### Import length bins & intervals
+        length_intervals = self.biology[ 'distributions' ][ 'length' ][ 'length_interval_arr' ]
+        
+        ### Age bins
+        age_intervals = self.biology[ 'distributions' ][ 'age' ][ 'age_interval_arr' ]
+ 
+        ### Process `haul_catch_df`
+        # Remove haul numbers not found within `length_df_copy`
+        haul_catch_filtered = (
+            haul_catch_df
+            .loc[ lambda x: x.haul_num.isin( length_df_copy.haul_num ) ]
+        )
+
+        ### Calculate station weights -- stratified by strata
+        weight_strata_station = (
+            pd.concat( [
+                haul_catch_filtered 
+                .groupby( 'stratum_num' )
+                .apply( lambda df: pd.Series( { 'stratum_weight': df.haul_weight.sum( ) } ) )
+                .reset_index( )
+                .assign( station = 1 ) , 
+                specimen_df_copy
+                .groupby( [ 'stratum_num' ] )
+                .apply( lambda df: pd.Series( { 'stratum_weight': df.weight.sum( ) } ) )
+                .reset_index()
+                .assign( station = 2 )
+            ] )
+        )
```

Comment:
It is a matter of style, but I think the code would be more readable if this is broken into 3 pieces:
```python
haul_catch_filtered["stratum_weight"] = ...
specimen_df_copy["stratum_weight"] = ...
weight_strata_station = pd.concat( [haul_catch_filtered, specimen_df_copy])
```
Not sure if I get the intention correct though, because I could not evaluate this.

Reply:
And there's virtue in vectoring operations: https://stackoverflow.com/a/66364784